staticd/static_bfd.c

```diff
@@ -0,0 +1,504 @@
+/*
+ * Static daemon BFD integration.
+ *
+ * Copyright (C) 2020 Network Device Education Foundation, Inc. ("NetDEF")
+ *                    Rafael F. Zalamena
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301 USA.
+ */
+
+#include <zebra.h>
+
+#include "lib/printfrr.h"
+#include "lib/srcdest_table.h"
+
+#include "staticd/static_routes.h"
+#include "staticd/static_zebra.h"
+
+#include "lib/openbsd-queue.h"
+
+DEFINE_MTYPE_STATIC(STATIC, ROUTE_GROUP, "route group memory usage");
+DEFINE_MTYPE_STATIC(STATIC, ROUTE_GROUP_MEMBER,
+		    "route group member memory usage");
+
+/** Global static BFD integration variables. */
+struct static_bfd_global {
+	/** Static route group list. */
+	struct srglist sbg_srglist;
+};
+
+static struct static_bfd_global sbglobal;
+
+/*
+ * Next hop BFD monitoring settings.
+ */
+static void static_next_hop_bfd_change(struct static_nexthop *sn,
+				       const struct bfd_session_status *bss)
+{
+	switch (bss->state) {
+	case BSS_UNKNOWN:
+		/* FALLTHROUGH: no known state yet. */
+	case BSS_ADMIN_DOWN:
+		/* NOTHING: we or the remote end administratively shutdown. */
+		break;
+	case BSS_DOWN:
+		/* Peer went down, remove this next hop. */
+		sn->path_down = true;
+		static_zebra_route_add(sn->pn, true);
```

Comment:
static_zebra_route_add(sn->pn, false); ?

Reply:
it has to be set to true.
if there are other multipath nexthops, then this routine installs other next hops.
this can lead to install floating routes ( routes with nexthop with distance bigger)

Reply:
true seems to be corret